Implement publisher list v4

zenparsing opened the pull request:
<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/9502

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [x] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

This PR modifies how publisher status and banner information is retrieved and stored on the client. Instead of downloading the entire publisher list dataset and storing that dataset in the rewards database, we will now download a small "index" of publishers and then download publisher data for individual publishers as-needed. There should be no user-visible changes, but existing test cases that involve publisher statuses or banner data should be checked for regressions.

### Basic functionality test

Flags:

```
--rewards=staging=true --enable-logging=stderr --v=0 --vmodule='server_publisher_fetcher=1,publisher_prefix_list_updater=1'
```

- Start browser with a clean profile using the flags above (additional `vmodule` entries can be used).
- Enable rewards.
- Dismiss the grant claim notification by clicking the small "x" in the rewards panel.
- Navigate to "https://laurenwags.github.io".
  - Verify that a "verified" check mark appears on the rewards badge.
- Open the rewards panel and click "Send a Tip".
  - Verify that the tip banner displays the correct information.


## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_info.cc:
```diff
@@ -105,77 +138,53 @@ bool DatabaseServerPublisherInfo::MigrateToV15(
-    callback(ledger::Result::LEDGER_OK);
-    return;
+  // TODO(zenparsing): Do we need to vacuum after this? Vacuum
+  // took 276ms and reduced the size from 137MB to 16MB.
```
@NejcZdovc After emptying the data from `server_publisher_info`, I think we need to vacuum to release disk space. Do we currently do any vacuuming?

NejcZdovc replied:
Would be great to add vacuum type in DB transaction so we can use it different situations

zenparsing replied:
I added a `VACUUM` DB command type. I decided to include it as part of this PR since only this PR exercises it.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,307 @@
+
+    // TODO(zenparsing): The previous JSON data had an "excluded"
+    // field, whereas the protobuf format does not. Do we still
+    // need this field?
```
@NejcZdovc See the comment above. The `ChannelResponse` message doesn't contain the equivalent of an "excluded" field. Do we still need this field?

zenparsing replied:
It looks like "excluded" is used in `SaveVisitInternal` when recording activity for AC. Do we not need that logic anymore? Or do I need to request that the "excluded" field added to the new protocol?

NejcZdovc replied:
afaik we will have different list for excluded list @yachtcaptain23 to confirm

yachtcaptain23 replied:
We don't use the excluded list anymore. cc: @evq

zenparsing replied:
@NejcZdovc should I remove the "excluded" field from the table and associated mojom struct?

NejcZdovc replied:
@yachtcaptain23 how come that we don't have excluded list anymore? is everybody aware of this change? cc @mandar-brave

NejcZdovc replied:
@zenparsing yes remove excluded field and all related logic for this flow. Note that we still need excluded from user flow

zenparsing replied:
Removed `excluded` from `server_publisher_info` table and mojom structs.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,307 @@
+}
+
+// TODO(zenparsing): This copied from components/brave_private_cdn.
+// Can we include it from bat-native-ledger?
```
@NejcZdovc This functionality is duplicated from `components/brave_private_cdn`. Are we "allowed" to pull in that code?

NejcZdovc replied:
@bridiver @kylehickinson  wdyt? see any problems?

kylehickinson replied:
I can't see where this was anymore (seems like the comment is gone?). Also I can't find a  `brave_private_cdn` in brave-core to check

zenparsing replied:
@kylehickinson [see here](https://github.com/brave/brave-core/blob/master/components/brave_private_cdn/private_cdn_helper.cc#L12) for the existing implementation.

The one in this PR is [here](https://github.com/brave/brave-core/pull/5568/files#diff-f53b2a5a835cc256a097da16fe4628e9R119). (I modified my version to avoid an unnecessary string copy.)

kylehickinson replied:
@zenparsing Based on [its BUILD.gn file](https://github.com/brave/brave-core/blob/master/components/brave_private_cdn/BUILD.gn) I dont see any issue with adding it to iOS (We already have a dep on `//base` anyways)

zenparsing replied:
I'll need to create an overload of `RemovePadding` that takes a `StringPiece`; I'll use that code once that change is in place.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc:
```diff
@@ -0,0 +1,142 @@
+  seconds <<= retry_count_;
+  retry_count_ += 1;
+  if (seconds > kMaxRetryAfterFailureDelay) {
+    seconds = kMaxRetryAfterFailureDelay;
```
Review from a privacy perspective.

bridiver commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc:
```diff
@@ -0,0 +1,160 @@
+
+namespace braveledger_publisher {
+
+PublisherListUpdater::PublisherListUpdater(bat_ledger::LedgerImpl* ledger)
```
is there a specific reason we can't use the component updater for this? We are trying to standardize on that across all platforms

zenparsing replied:
We'll move to component updater in a future PR.

bridiver commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_reader.cc:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  std::string uncompressed;
+  switch (message.compression_type()) {
```
why are we compressing data inside the publisher list instead of just compressing the entire response payload?

zenparsing replied:
This design is inspired by Safe Browsing, and will give us a lot of flexibility for different compression approaches in the future (similar to how Safe Browsing uses Rice encoding).

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/common/time_util.h:
```diff
@@ -25,6 +25,13 @@ uint32_t GetYear(const base::Time& time);
 
+base::TimeDelta GetRandomizedDelay(base::TimeDelta delay);
+
+base::TimeDelta GetRandomizedDelayWithBackoff(
```
These helpers are used in `publisher_list_updater.cc` and will be used as we refactor timers to use the  timer classes from `base`.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_publisher_list.cc:
```diff
@@ -0,0 +1,194 @@
+const char kTableName[] = "publisher_list";
+
+constexpr size_t kHashPrefixSize = 4;
+constexpr size_t kMaxInsertRecords = 100'000;
```
In order to avoid sending the database a SQL query of unbounded length when we insert hash prefixes into the publisher list table, we divide up the work into chunks into a series of inserts with a bounded number of values.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc:
```diff
@@ -1155,26 +1155,53 @@ void LedgerImpl::DeleteActivityInfo(
+    ledger::ServerPublisherInfoPtr server_info,
+    const std::string& publisher_key,
+    ledger::GetServerPublisherInfoCallback callback) {
+  if (bat_publisher_->ShouldFetchServerPublisherInfo(server_info.get())) {
```
After getting the data from the database layer, we check if the record is expired, and if so, re-fetch it from the private CDN.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc:
```diff
@@ -140,17 +142,18 @@ void Publisher::SaveVisit(
+      this, _1, publisher_key, visit_data, duration, window_id, callback);
 
-  ledger_->GetServerPublisherInfo(publisher_key, server_callback);
+  ledger_->SearchPublisherList(
```
On page visits, we want to search the publisher prefix list before attempting to call `GetServerPublisherInfo` (which could trigger a fetch).

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc:
```diff
@@ -264,21 +263,15 @@ void Publisher::SaveVisitInternal(
-      publisher_info->id,
-      server_excluded,
-      static_cast<ledger::PublisherExclude>(publisher_info->excluded));
+  bool excluded =
```
The `excluded` field is not included in channel responses and has been removed from `ServerPublisherInfo` structs and tables. It is still available in `PublisherInfo`.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_reader.cc:
```diff
@@ -41,6 +42,17 @@ PublisherListReader::ParseError PublisherListReader::Parse(
       uncompressed = std::move(*message.mutable_prefixes());
       break;
     }
+    case PublisherList::BROTLI_COMPRESSION: {
```
Brotli gives better results that gzip for the sorted prefix list, and is easy to support since we are already using it for the channel response payload. In the future, we may introduce more specialized compression algorithms similar to how Safe Browsing uses Rice encoding.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.h:
```diff
@@ -0,0 +1,67 @@
+
+// Automatically updates the publisher prefix list store on regular
+// intervals.
+class PublisherListUpdater {
```
In the future, and likely as a blocker for "rewards everywhere", we will transition to distributing the publisher prefix list using component updater. That effort will require some architectural work in the rewards service.

NejcZdovc replied:
can you please log an issue for it, so that we don't miss it

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_status_helper.h:
```diff
@@ -0,0 +1,32 @@
+namespace braveledger_publisher {
+
+// Refreshes the publisher status for each entry in the specified list
+void RefreshPublisherStatus(
```
These functions are used when a list of records is returned from the database that includes the publisher status. For those queries, the `server_publisher_info` table is linked by left join. Now that we are using the `server_publisher_info` table as a cache, it may be missing data, or it may have outdated data.

These functions are used by queries that need up-to-date publisher status information to update the publisher status for each entry in the list before calling the supplied callback.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,317 @@
+  return nullptr;
+}
+
+bool RemovePadding(base::StringPiece* padded_string) {
```
This can be removed after we add a `StringPiece` overload for the `RemovePadding` method in `components/brave_private_cdn/` (#5791).

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,317 @@
+  CallbackVector& callbacks = callback_map_[publisher_key];
+  callbacks.push_back(callback);
+  if (callbacks.size() > 1) {
+    BLOG(1, "Fetch already in progress for publisher " << publisher_key);
```
The `Fetch` method can be called from various places that want updated publisher data, so it's important to dedupe any requests that come in for the same publisher key.

emerick commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_status_helper.cc:
```diff
@@ -0,0 +1,135 @@
+    ledger::PendingContributionInfoList&& info_list,
+    ledger::PendingContributionInfoListCallback callback) {
+  PublisherStatusMap map;
+  for (auto& info : info_list) {
```
Can `info` be const?

emerick commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_status_helper.cc:
```diff
@@ -0,0 +1,135 @@
+    ledger::PublisherInfoList&& info_list,
+    ledger::PublisherInfoListCallback callback) {
+  PublisherStatusMap map;
+  for (auto& info : info_list) {
```
Can `info` be const?

emerick commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,317 @@
+ledger::ServerPublisherInfoPtr ServerPublisherInfoFromMessage(
+    const publishers_pb::ChannelResponseList& message,
+    const std::string& expected_key) {
+  for (auto& entry : message.channel_responses()) {
```
Can `entry` be const?

emerick commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,317 @@
+      [this, publisher_key, shared_info](ledger::Result result) {
+        if (result != ledger::Result::LEDGER_OK) {
+          BLOG(0, "Error saving server publisher info record");
+        }
```
Just checking: We should still call `RunCallbacks` in this scenario?

fmarier commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/channel_response.proto:
```diff
@@ -0,0 +1,43 @@
+
+package publishers_pb;
+
+message SocialLinks {
```
Are GitHub and Reddit links handled as URLs?

zenparsing replied:
Yes, they are either URLs or empty strings.

fmarier commented on vendor/bat-native-ledger/src/bat/ledger/internal/static_values.h:
```diff
@@ -25,11 +25,11 @@
 
 #define PUBLISHER_STAGING_SERVER     \
-"https://publishers-staging-distro.basicattentiontoken.org"
+"https://pcdn.brave.software"
```
Staging should be `pcdn.bravesoftware.com`.

`pcdn.brave.software` is dev, as correctly defined below.

yachtcaptain23 replied:
@fmarier I'm getting an error fetching from `pcdn.bravesoftware.com`. I believe devops only setup production and dev

fmarier replied:
Should be fixed now: https://bravesoftware.slack.com/archives/C011BLDBB63/p1591895026073400

zenparsing replied:
Fixed

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_publisher_list.cc:
```diff
@@ -0,0 +1,194 @@
+    callback(ledger::Result::LEDGER_ERROR);
+    return;
+  }
+  if (reader->size() == 0) {
```
I think I'd like to leave this as-is, and maybe add an `empty` method to that class later. OK with you?

emerick replied:
No problem.

zenparsing commented:
Tasks from 2020-06-17 meeting:

- Set "last fetch" timestamp *after* a complete insert into the database. **(done)**
  - As noted in comments, the "last fetch" timestamp is recorded even if the database insert failed, in order to avoid overloading server on client failure. Feel free to comment on this.
- Consider renaming "publisher_list" files and classes. **(in progress)**
- Move "chrome://..." URL prefixes into front end. **(deferred)**
  - I recommend this be addressed in a future PR, in order to have a clean history and clear test plan.
- `RefreshPublisher` should not be called from the status update helper. Call `GetServerPublisherInfo` instead. **(done)**
- Consider adding `ContributeUnverifiedPublisher(publisher_key)`. **(deferred)**
  - I recommend addressing this in a follow up, as the addition will not be trivial.
- Review lambda usage. **(done)**
  - The lambdas usage looks readable to me (when compared to `std::bind`), but will be happy to change in response to PR reviews.
- Rebase and update v27 test database file. **(done)**

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc:
```diff
@@ -0,0 +1,159 @@
+  timer_.Stop();
+}
+
+void PublisherListUpdater::SetOnPublisherListUpdatedCallback(
```
could you tell a little bit more about this one?

zenparsing replied:
I modified it slightly so that `StartAutoUpdate` takes a callback instead of having this separate method, but the idea is that, ideally,`PublisherPrefixListUpdater` shouldn't have to know about the kinds of things that we might want to do after an update; it should just know how to perform the update, and let callers do what they want after each update (there are two callouts that happen in `publisher.cc`).

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc:
```diff
@@ -0,0 +1,159 @@
+    // optimistically assume that it is a server issue and retry
+    // with back-off.
+    BLOG(0, "Failed to parse publisher prefix list: "
+        << static_cast<int>(parse_error));
```
cast is not needed

zenparsing replied:
The cast is needed since `ParseError` is an enum class.

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc:
```diff
@@ -0,0 +1,159 @@
+    return;
+  }
+
+  if (reader->size() == 0) {
```
use empty

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc:
```diff
@@ -0,0 +1,159 @@
+  }
+
+  if (reader->size() == 0) {
+    BLOG(0, "Publisher prefix list did not contain any values");
```
change this to level 1

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc:
```diff
@@ -0,0 +1,159 @@
+}
+
+void PublisherListUpdater::OnFetchCompleted(
+    const ledger::UrlResponse& response) {
```
add `BLOG(7, ledger::UrlResponseToString(__func__, response));`

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc:
```diff
@@ -0,0 +1,159 @@
+void PublisherListUpdater::StartFetchTimer(
+    const base::Location& posted_from,
+    base::TimeDelta delay) {
+  BLOG(1, "Scheduling publisher prefix list update in "
```
we should have check for rewards main toggle so that we exit if rewards is off

zenparsing replied:
This function won't get called if rewards is off, because `StopAutoUpdate` will have been called. Can we leave it as-is for now, so that the updater doesn't have to know about "rewards enabled" and callers can just use `StartAutoUpdate`/`StopAutoUpdate`?

NejcZdovc replied:
sure

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.h:
```diff
@@ -0,0 +1,67 @@
+
+  bat_ledger::LedgerImpl* ledger_;  // NOT OWNED
+  base::OneShotTimer timer_;
+  bool auto_update_ = false;
```
not needed if we add rewards enabled check in start timer

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.h:
```diff
@@ -0,0 +1,67 @@
+  // Cancels the auto updater
+  void StopAutoUpdate();
+
+  using OnPublisherListUpdatedCallback = std::function<void()>;
```
please move it out of class

zenparsing replied:
done.

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.h:
```diff
@@ -0,0 +1,67 @@
+ private:
+  void StartFetchTimer(
+      const base::Location& posted_from,
+      base::TimeDelta delay);
```
`const base::TimeDelta&`

zenparsing replied:
According to the [code comments](https://cs.chromium.org/chromium/src/base/time/time.h?q=timedelta&sq=package:chromium&g=0&l=24) in `time.h`, we should prefer passing `TimeDelta` by value.

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,317 @@
+      return ledger::PublisherStatus::VERIFIED;
+    case publishers_pb::UPHOLD_ACCOUNT_NO_KYC:
+      return ledger::PublisherStatus::CONNECTED;
+    default:
```
let's not use default as then we can miss if new ones are added

zenparsing replied:
Unfortunately, for protobuf generated code, we get compile errors if we leave off the default.

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,317 @@
+
+ledger::ServerPublisherInfoPtr ServerPublisherInfoFromMessage(
+    const publishers_pb::ChannelResponseList& message,
+    const std::string& expected_key) {
```
add check for expected_key being empty

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,317 @@
+
+void ServerPublisherFetcher::OnFetchCompleted(
+    const std::string& publisher_key,
+    const ledger::UrlResponse& response) {
```
add `BLOG(6, ledger::UrlResponseToString(__func__, response));`

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc:
```diff
@@ -0,0 +1,317 @@
+  auto server_info = ledger::ServerPublisherInfo::New();
+  server_info->publisher_key = publisher_key;
+  server_info->status = ledger::PublisherStatus::NOT_VERIFIED;
+  server_info->updated_at =
```
use GetCurrentTimeStamp from time util

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.h:
```diff
@@ -0,0 +1,78 @@
+  void PurgeExpiredRecords();
+
+ private:
+  using CallbackVector =
```
move out of class

zenparsing replied:
Since this alias is only used internally in this class, there's not much benefit in moving it to `ledger.h`. We could also just use `std::vector<GetServerPublisherInfoCallback>` directly if we would prefer not to have an alias here.

NejcZdovc replied:
I was thinking more info moving it to line 22 for example in this file, not in `ledger.h`

zenparsing replied:
done.

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.h:
```diff
@@ -0,0 +1,78 @@
+
+  // Returns a value indicating whether the specified server info
+  // is expired
+  bool IsExpired(ledger::ServerPublisherInfo* server_info);
```
see comment above

zenparsing replied:
done.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/common/time_util.cc:
```diff
@@ -33,4 +35,23 @@ uint64_t GetCurrentTimeStamp() {
+    base::TimeDelta delay,
+    base::TimeDelta max_delay,
+    int backoff_count) {
+  for (int i = 0; i < backoff_count; ++i) {
+    delay *= 2;
+    if (delay > max_delay) {
+      delay = max_delay;
+      break;
+    }
+  }
```
`delay` here is a `TimeDelta`, so unfortunately that won't work. I recommend that we try to use `TimeDelta`s where possible, since they have things like overflow checking built-in, and their methods have readable names.

As far as this loop goes, I suspect that we'll have other problems before `backoff_count` gets so large that performance becomes an issue.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database.cc:
```diff
@@ -477,6 +483,12 @@ void Database::GetServerPublisherInfo(
 }
 
+void Database::DeleteExpiredServerPublisherInfo(
+    int64_t max_age_seconds,
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_amounts.cc:
```diff
@@ -179,57 +182,59 @@ bool DatabaseServerPublisherAmounts::MigrateToV15(
+void DatabaseServerPublisherAmounts::InsertOrUpdate(
+    ledger::DBTransaction* transaction,
+    const ledger::ServerPublisherInfo& server_info) {
+  DCHECK(transaction);
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_info.cc:
```diff
@@ -57,6 +58,34 @@ bool DatabaseServerPublisherInfo::CreateIndexV7(
+bool DatabaseServerPublisherInfo::CreateIndexV28(
+    ledger::DBTransaction* transaction) {
+  DCHECK(transaction);
+  return InsertIndex(transaction, kTableName, "publisher_key");
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_info.cc:
```diff
@@ -57,6 +58,34 @@ bool DatabaseServerPublisherInfo::CreateIndexV7(
+  const std::string query = base::StringPrintf(
+      "CREATE TABLE %s "
+      "("
+      "publisher_key LONGVARCHAR PRIMARY KEY NOT NULL UNIQUE,"
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_info.cc:
```diff
@@ -108,78 +140,64 @@ bool DatabaseServerPublisherInfo::MigrateToV15(
-    const std::vector<ledger::PublisherBanner>& list,
-    ledger::ResultCallback callback) {
-  banner_->InsertOrUpdateList(list, callback);
+  ledger_->RunDBTransaction(std::move(transaction),
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_info.cc:
```diff
@@ -264,11 +282,79 @@ void DatabaseServerPublisherInfo::OnGetRecord(
+    int64_t max_age_seconds,
+    ledger::ResultCallback callback) {
+  int64_t cutoff =
+      static_cast<int64_t>(base::Time::Now().ToDoubleT()) - max_age_seconds;
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_links.cc:
```diff
@@ -182,62 +185,66 @@ bool DatabaseServerPublisherLinks::MigrateToV15(
-    const std::vector<ledger::PublisherBanner>& list) {
+    const ledger::ServerPublisherInfo& server_info) {
   DCHECK(transaction);
+  DCHECK(!server_info.publisher_key.empty());
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_list.cc:
```diff
@@ -0,0 +1,197 @@
+constexpr size_t kHashPrefixSize = 4;
+constexpr size_t kMaxInsertRecords = 100'000;
+
+void DropAndCreateTableV28(ledger::DBTransaction* transaction) {
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_list.cc:
```diff
@@ -0,0 +1,197 @@
+    NOTREACHED();
+  }
+
+  auto command = ledger::DBCommand::New();
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_list.cc:
```diff
@@ -0,0 +1,197 @@
+  DCHECK(transaction);
+  switch (target) {
+    case 28:
+      MigrateToV28(transaction);
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_list.cc:
```diff
@@ -0,0 +1,197 @@
+
+bool DatabaseServerPublisherList::Migrate(
+    ledger::DBTransaction* transaction,
+    int target) {
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_list.cc:
```diff
@@ -0,0 +1,197 @@
+    std::unique_ptr<braveledger_publisher::PublisherListReader> reader,
+    ledger::ResultCallback callback) {
+  if (reader_) {
+    BLOG(0, "Publisher prefix list batch insert in progress");
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_list.cc:
```diff
@@ -0,0 +1,197 @@
+    callback(ledger::Result::LEDGER_ERROR);
+    return;
+  }
+  if (reader->size() == 0) {
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_list.cc:
```diff
@@ -0,0 +1,197 @@
+       ++count, ++iter) {
+    auto prefix = *iter;
+    DCHECK(prefix.size() >= kHashPrefixSize);
+    values.append(iter == begin ? "(x'" : "'),(x'");
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_list.h:
```diff
@@ -0,0 +1,44 @@
+  explicit DatabaseServerPublisherList(bat_ledger::LedgerImpl* ledger);
+  ~DatabaseServerPublisherList() override;
+
+  bool Migrate(ledger::DBTransaction* transaction, int target) override;
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc:
```diff
@@ -1149,26 +1149,53 @@ void LedgerImpl::DeleteActivityInfo(
+}
+
+void LedgerImpl::DeleteExpiredServerPublisherInfo(
+    int64_t max_age_seconds,
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc:
```diff
@@ -24,63 +26,61 @@ using std::placeholders::_2;
+  // for the specified publisher.
+  server_publisher_fetcher_->Fetch(publisher_key,
+      [this, callback](auto server_info) {
+        auto status = server_info
```
I believe it should be checked as part of the conditional expression?

NejcZdovc replied:
not sure that I understand what you  mean

zenparsing replied:
It's checked as part of the expression `server_info ? server_info->status : ledger::PublisherStatus::NOT_VERIFIED`

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_list.cc:
```diff
@@ -0,0 +1,197 @@
+void DatabaseServerPublisherList::InsertNext(
+    PrefixIterator begin,
+    ledger::ResultCallback callback) {
+  DCHECK(reader_);
```
done

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/request/request_publisher.cc:
```diff
@@ -3,19 +3,34 @@
-      "/api/v3/public/channels?page=%d",
-      page);
+std::string GetPublisherListUrl() {
+  return BuildUrl(
```
This will need to use a different server type from the private CDN.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/static_values.h:
```diff
@@ -25,11 +25,11 @@
+"https://rewards.brave.com"
 #define PUBLISHER_DEVELOPMENT_SERVER  \
-"https://creators-distro.brave.software"
+"https://rewards-stg.bravesoftware.com"
```
This still needs to be updated. cc @yachtcaptain23

NejcZdovc replied:
what is status on this one @zenparsing @yachtcaptain23 ?

yachtcaptain23 replied:
Unless I'm mistaken, this is already implemented server-side and client-side. Recommend resolving.

zenparsing replied:
@yachtcaptain23 Just to confirm again - we're going to use the same prefix list URL for both staging and development, correct?

NejcZdovc replied:
not sure how we can use the same list if publishers are different

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc:
```diff
@@ -24,63 +26,61 @@ using std::placeholders::_2;
-          callback));
+bool Publisher::ShouldFetchServerPublisherInfo(
+    ledger::ServerPublisherInfo* server_info) {
+  return server_publisher_fetcher_->IsExpired(server_info);
```
done.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.h:
```diff
@@ -20,16 +21,23 @@ class LedgerImpl;
 
-  // Called when timer is triggered
-  void OnTimer(uint32_t timer_id);
+  bool ShouldFetchServerPublisherInfo(base::Time last_updated_time);
```
done.

zenparsing commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc:
```diff
@@ -24,63 +26,61 @@ using std::placeholders::_2;
-  if (!rewards_enabled) {
-    server_list_->ClearTimer();
-    return;
+  if (rewards_enabled) {
```
Can this be a follow-up?

NejcZdovc replied:
sure, please file an issue for it

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_publisher_prefix_list.cc:
```diff
@@ -0,0 +1,204 @@
+constexpr size_t kHashPrefixSize = 4;
+constexpr size_t kMaxInsertRecords = 100'000;
+
+std::tuple<PrefixIterator, std::string, size_t> GetPrefixInsertList(
```
can you please add unit tests for this one

zenparsing replied:
Since this is internal, I chose to add a unit test for the "Reset" method, which tests how large inserts are split up between transactions.

NejcZdovc replied:
Sounds good

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_publisher_prefix_list.cc:
```diff
@@ -0,0 +1,204 @@
+
+  ledger_->RunDBTransaction(std::move(transaction), [callback](
+      ledger::DBCommandResponsePtr response) {
+    if (response && response->result) {
```
we should also check for `response->status != ledger::DBCommandResponse::Status::RESPONSE_OK`

zenparsing replied:
done.

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_publisher_prefix_list.cc:
```diff
@@ -0,0 +1,204 @@
+  ledger_->RunDBTransaction(std::move(transaction), [callback](
+      ledger::DBCommandResponsePtr response) {
+    if (response && response->result) {
+      for (const auto& record : response->result->get_records()) {
```
can we just do `response->result->get_records()[0].get();` so that we don't need loop?

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_info.cc:
```diff
@@ -6,6 +6,8 @@
 #include <utility>
 
 #include "base/strings/stringprintf.h"
+#include "base/time/time.h"
```
not used

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/database/database_server_publisher_info.h:
```diff
@@ -22,25 +22,19 @@ class DatabaseServerPublisherInfo: public DatabaseTable {
-      ledger::DBTransaction* transaction,
-      ledger::ServerPublisherInfoPtr info);
+  void DeleteExpiredRecords(
+      int64_t max_age_seconds,
```
nit: `const`

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h:
```diff
@@ -417,19 +419,27 @@ class LedgerImpl : public ledger::Ledger {
+      ledger::GetServerPublisherInfoCallback callback);
+
+  void DeleteExpiredServerPublisherInfo(
+      int64_t max_age_seconds,
```
nit: `const`

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.h:
```diff
@@ -14,6 +14,7 @@
 #include <vector>
 
 #include "base/memory/scoped_refptr.h"
+#include "base/time/time.h"
```
not used

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/request/request_util.h:
```diff
@@ -18,7 +18,8 @@ enum class ServerTypes {
   kPayments,
-  kAPI
+  kAPI,
+  kPrivateCdn
```
should this be `kPrivateCDN`? as it's short version

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/request/request_publisher.cc:
```diff
@@ -3,19 +3,34 @@
 
+namespace {
+
+constexpr char kPrefix[] = "";
```
I would say that this can be removed, based on the fact that @yachtcaptain23 said that we will not do versioning, but protobuff will handle it for us

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_status_helper.cc:
```diff
@@ -0,0 +1,136 @@
+
+void RefreshNext(std::shared_ptr<RefreshTaskInfo> task_info) {
+  // Find the first map element that has an expired status.
+  task_info->current = std::find_if(
```
check if `task_info` is null

zenparsing replied:
I've added a DCHECK since it should never be empty.

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_status_helper.cc:
```diff
@@ -0,0 +1,136 @@
+    bat_ledger::LedgerImpl* ledger,
+    PublisherStatusMap&& status_map,
+    std::function<void(PublisherStatusMap)> callback) {
+  RefreshNext(std::make_shared<RefreshTaskInfo>(
```
DCHECK(ledger)

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_status_helper.cc:
```diff
@@ -0,0 +1,136 @@
+    bat_ledger::LedgerImpl* ledger,
+    ledger::PublisherInfoList&& info_list,
+    ledger::PublisherInfoListCallback callback) {
+  PublisherStatusMap map;
```
DCHECK(ledger)

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_status_helper.cc:
```diff
@@ -0,0 +1,136 @@
+    bat_ledger::LedgerImpl* ledger,
+    ledger::PendingContributionInfoList&& info_list,
+    ledger::PendingContributionInfoListCallback callback) {
+  PublisherStatusMap map;
```
DCHECK(ledger)

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.h:
```diff
@@ -12,6 +12,7 @@
 #include <vector>
 
 #include "base/gtest_prod_util.h"
+#include "base/time/time.h"
```
not used

zenparsing replied:
done

NejcZdovc commented on vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_prefix_list_updater.cc:
```diff
@@ -0,0 +1,161 @@
+}
+
+void PublisherPrefixListUpdater::OnPrefixListInserted(
+    ledger::Result result) {
```
nit: `const`

zenparsing replied:
done

zenparsing commented:
- Updated `rewards_database.cc` so that VACUUM is not performed if transaction could not be committed.
- Updated `rewards_database.cc` a failed VACUUM is logged but not returned as an error to the caller.
- Updated `database_server_publisher_*.cc` delete methods to no-op when expired records list is empty.

zenparsing commented:
CI failed due to `audit_deps` and an unrelated macos `gsync` failure (macos verified locally).
